src/app/api/label-mapping/label-mapping.service.ts

```diff
@@ -0,0 +1,168 @@
+    const baseLabel = labelledEntity?.substring(0, 3).toUpperCase()
+
+    //find the latest updated LabelMapping for the given labelledEntity
+    const existingLabelWithCurrentLabelEntity = await this.db.labelMapping.findFirst({
+      where: {
+        labelledEntity,
+      },
+      orderBy: {
+        createdAt: 'desc',
+      },
+    })
+
+    //find all the LabelMapping that starts with the baseLabel substring
+    const existingLabelWithSameBaseLabel = await this.db.labelMapping.findMany({
+      where: {
+        label: {
+          startsWith: baseLabel,
+        },
+      },
+      orderBy: {
+        createdAt: 'desc',
+      },
+    })
+
+    //case I -> if no items are found with the baseLabel. In this case, existingLabelWithSameBaseLabel is also undefined
+    if (existingLabelWithSameBaseLabel.length === 0) {
+      const label = `${baseLabel}-001`
+      await this.insertLabelMapping(label, labelledEntity)
+      return label
+    }
+
+    //case II -> if existingLabelWithCurrentLabelEntity is undefined but there are existingLabelWithSameBaseLabel
+    if (!existingLabelWithCurrentLabelEntity && existingLabelWithSameBaseLabel.length > 0) {
+      //suffixNumber is the number assigned to conflicting labels. for ex: OUT2-001 where 2 is suffix number
+      //The line below finds the highest suffix number from the existingLabelWithCurrentLabelEntity. If no suffixNumber is found then it returns null.
+      const suffixNumber = this.extractLabelSuffixNumber(
+        this.findLabelMappingWithHighestSuffix(existingLabelWithSameBaseLabel).label,
+      )
+      const currentBaseLabel = `${baseLabel}${suffixNumber ? Number(suffixNumber) + 1 : '2'}`
+      const newLabel = `${currentBaseLabel}-001`
+      await this.insertLabelMapping(newLabel, labelledEntity)
+      return newLabel
+    }
+
+    //case III -> if both existingLabelWithSameBaseLabel and existingLabelWithCurrentLabelEntity are truthy
+    if (existingLabelWithCurrentLabelEntity && existingLabelWithSameBaseLabel.length > 0) {
+      const newLabel = `${existingLabelWithCurrentLabelEntity.label.split('-')[0]}-${this.getNextLabelCount(existingLabelWithCurrentLabelEntity.label)}`
+      await this.insertLabelMapping(newLabel, labelledEntity)
+      return newLabel
+    }
+  }
+
+  /**
+   * This method finds the LabelMapping with the highest suffix.
+   * ex: [{label: "COP1-001"}, {label: "COP2-002"}] will return {label: "COP2-002"} where 2 is the highest suffix
+   * @param labels PrismaLabelMapping[]
+   * @returns PrismaLabelMapping
+   */
+  private findLabelMappingWithHighestSuffix(labels: PrismaLabelMapping[]) {
+    return labels.reduce((maxLabel, currentLabel) => {
+      const maxSuffix = this.extractLabelSuffixNumber(maxLabel.label) || 0
+      const currentSuffix = this.extractLabelSuffixNumber(currentLabel.label) || 0
+      return currentSuffix > maxSuffix ? currentLabel : maxLabel
+    }, labels[0])
```

Comment:
I think this could be better done using an SQL query rather than this way - it would be much more scalable.

```sql
SELECT *
FROM LabelMappings
WHERE CAST(SUBSTR(label, 5) AS UNSIGNED) = (
    SELECT MAX(CAST(SUBSTR(label, 5) AS UNSIGNED))
    FROM LabelMappings
);
```

Reply:
The suggested query is not working and giving unintended errors. Please kindly suggest a working query so that I can address the change.